pytables.ipynb

```diff
@@ -0,0 +1,287 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "import tables\n",
+    "\n",
+    "import pandas as pd\n",
+    "from neo4j.v1 import GraphDatabase\n",
+    "import hetio.readwrite\n",
+    "import hetio.neo4j\n",
+    "import hetio.pathtools\n",
+    "import numpy as np\n",
+    "from scipy import sparse\n",
+    "import time\n",
+    "from hetmech.matrix import get_node_to_position, metaedge_to_adjacency_matrix"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "url = 'https://github.com/dhimmel/hetionet/raw/76550e6c93fbe92124edc71725e8c7dd4ca8b1f5/hetnet/json/hetionet-v1.0.json.bz2'\n",
+    "graph = hetio.readwrite.read_graph(url)\n",
+    "metagraph = graph.metagraph"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# NPY Saving and loading"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 38,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "r, c, gig_adj = metaedge_to_adjacency_matrix(graph, 'GiG')\n",
+    "r, c, ctd_adj = metaedge_to_adjacency_matrix(graph, 'CtD')\n",
+    "r, c, sparse_gig_adj = metaedge_to_adjacency_matrix(graph, 'GiG', sparse_threshold=1)\n",
+    "r, c, sparse_ctd_adj = metaedge_to_adjacency_matrix(graph, 'CtD', sparse_threshold=1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Save matrices"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "930.0286769866943 ms\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Dense matrices into .npy\n",
+    "t1 = time.time()\n",
+    "np.save('data/gig_adj', gig_adj)\n",
+    "np.save('data/ctd_adj', ctd_adj)\n",
+    "t2 = time.time()\n",
+    "print(f'{(t2 - t1)*1000} ms')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 43,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "19.55270767211914 ms\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Sparse matrices into .npy\n",
+    "t1 = time.time()\n",
+    "np.save('data/sparse_gig_adj', sparse_gig_adj)\n",
+    "np.save('data/sparse_ctd_adj', sparse_ctd_adj)\n",
```

Comment:
Are we sure this is saving them as sparse (and not converting them to dense, then saving)? Should we be using [`scipy.sparse.save_npz`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.save_npz.html)?

Reply:
I just added a new commit which looks at this. It seems like `np.load` is faster than `sparse.load_npz`, and similarly `np.save` seems faster than `sparse.save_npz`.
One thing I found odd, which perhaps you could offer insight on is the following: when I first run (`Restart & Run All`), I get something like `25 ms` for the sparse load time from `.npz`. When I rerun the cell, the time invariable drops to more like `5 ms`. There must be some kind of caching going on here, right?

Reply:
That does sound fishy, @zietzm , and caching of some sort would be my first thought.

Try running the cell, then running a bunch of other cells (to fill up cache), and running the sparse load cell again, to test the hypothesis?

Reply:
Strange. After working on some other things for a while and running other cells, I retried it and was getting ~ `15 ms`, which wasn't changing much at all. I opened it 100 times and got an average speed of `12.8 ms` for the two matrices.

Reply:
What could explain this?

1. imports. The first time you run the loader, it may have to import some libraries. This only happens for the first time the modules are imported.

2. Swapping. If you're using all your RAM (memory), perhaps you end up having to free up memory or writing to the high-latency swap. These could both be slow.